ably/options.go

```diff
@@ -420,3 +381,309 @@ func (p *PaginateParams) EncodeValues(out *url.Values) error {
 	p.ScopeParams.EncodeValues(out)
 	return nil
 }
+
+type ClientOptions []func(*clientOptions)
+
+func NewClientOptions(key string) ClientOptions {
+	return ClientOptions{func(os *clientOptions) {
+		os.Key = key
+	}}
+}
+
+type AuthOptions []func(*authOptions)
+
+// A Tokener is or can be used to get a TokenDetails.
+type Tokener interface {
+	IsTokener()
+	isTokener()
```

Comment:
These methods seems not used anywhere `TokenRequest`, `TokenString` and `TokenRequest` implements these, but they are no op. And used internally meaning library users are not supposed to bring their own implementation. We probably need a private interface method., leaving `Tokener` accessible to library consumers with a limited set of options/implementations that are provided by the lib.

```
type Tokener interface {
	isToken()
}
```

will satisfy this requirement and ensure that library users can only pick between the three options we currently provide. I believe this will make the panic `panic(fmt.Errorf("unhandled TokenLike: %T", v))` redundant because it will never happenn.

Reply:
> We probably need a private interface method., leaving `Tokener` accessible to library consumers with a limited set of options/implementations that are provided by the lib.

That's exactly the purpose of `isToken`, but we also need a public method (`IsToken`) so that users of the library can look up which types (ie. `TokenRequest`, `TokenString` and `TokenRequest`) implement `Tokener`.

>  I believe this will make the panic `panic(fmt.Errorf("unhandled TokenLike: %T", v))` redundant because it will never happenn.

It is, it's just a check in case we add a new `Tokener` but forget to add the correspondent case to the switch.